core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -233,6 +233,43 @@ final class TMap[K, V] private (
       TExit.unit
     })
 
+  /**
+   * Removes bindings matching predicate and returns the removed entries.
+   */
+  def removeIfReturning(p: (K, V) => Boolean): USTM[Chunk[(K, V)]] =
+    new ZSTM((journal, _, _, _) => {
+      val f        = p.tupled
+      val buckets  = tBuckets.unsafeGet(journal)
+      val capacity = buckets.array.length
+
+      var i       = 0
+      var newSize = 0
+      var removed = Chunk[(K, V)]()
```

Comment:
Perhaps it makes sense to use `ChunkBuilder` instead.

Reply:
I used `ChunkBuilder` now. Note that it does change the order of the output (because the builder is appending and not prepending)

Reply:
Order isn't important here, so this should be fine.

Reply:
Well I don't think there is a deterministic order because the map is unordered but to the extent we did care about order isn't this more "correct" because we are returning the removed elements in the same order we originally traversed them?